File: python/data_structures/linked_list.py
# linked list will have a head, and a possibly a tail
# without at tail adding to tail will take O(n)
# add_to_head
# add_to_tail


class Node():
    def __init__(self, val):
        self.value = val
        self.next = None

    def set_next(self, node):
        self.next = node

    def __repr__(self):
        return str(self.value)


class LinkedList():
    def __init__(self, node=None):
        self.head = node
        self.tail = node

    def add_to_head(self, node):
        if self.head is None:
            self.tail = node
        node.next = self.head
        self.head = node

    def add_to_tail(self, node):
        if self.head is None:
            self.head = node
            self.tail = node
            return
        self.tail.next = node
        self.tail = node

    def remove_from_head(self):
        current_head = self.head
        self.head = self.head.next
        return current_head

    def remove_from_tail(self):
        current_tail = self.tail
        pre_tail = self.head
        while pre_tail.next is not current_tail:
            pre_tail = pre_tail.next
        self.tail = pre_tail
        self.tail.next = None
        return current_tail

    def __iter__(self):
        node = self.head
        while node is not None:
            yield node
            node = node.next

    def __repr__(self):
        nodes = []
        for node in self:
            nodes.append(str(node.value))

        return " <- ".join(nodes)


if __name__ == "__main__":
    linked_list = LinkedList()
    linked_list.add_to_head(Node(1))
    print(linked_list)
    linked_list.add_to_head(Node(2))
    print(linked_list)
    linked_list.add_to_tail(Node(3))
    linked_list.add_to_tail(Node(4))
    linked_list.add_to_tail(Node(5))
    linked_list.add_to_tail(Node(6))
    print(linked_list)
    print(f"removed head: {linked_list.remove_from_head()}")
    print(f"removed tail: {linked_list.remove_from_tail()}")
    print(linked_list)

File: js/data_structures/linked_list.ts
class Node<T> {
    value: T;
    next: Node<T> | null;

    constructor(value: T) {
        this.value = value
        this.next = null
    }

    set_next(node: Node<T>) {
        this.next = node
    }

    toString() {
        return this.value
    }
}

class LinkedList<T> {
    #head: Node<T> | null
    #tail: Node<T> | null

    constructor(node: Node<T> | null = null) {
        this.#head = node
        this.#tail = node
    }

    addToHead(node: Node<T>) {
        node.next = this.#head;
        this.#head = node
        if (!this.#tail) this.#tail = node

    }

    addToTail(node: Node<T>) {
        if (!this.#head) {
            this.#head = node
            this.#tail = node
            return
        }

        this.#tail.next = node
        this.#tail = node
    }

    removeFromHead() {
        const current_head = this.#head
        this.#head = this.#head.next;
        return current_head;
    }

    removeFromTail() {
        const tail = this.#tail;
        let pre_tail = this.#head;

        while (pre_tail.next !== tail) {
            pre_tail = pre_tail.next
        }

        this.#tail = pre_tail
        this.#tail.next = null
        return tail
    }

    *[Symbol.iterator]() {
        let current = this.#head;
        while (current) {
            yield current.value;
            current = current.next
        }
    }

    toString() {
        let nodes = []
        for (let node of this) {
            nodes.push(node);
        }

        return nodes.join(" <- ");
    }
}




const main = () => {
    const linkedList = new LinkedList<number>()
    linkedList.addToHead(new Node(1))
    linkedList.addToHead(new Node(2))
    linkedList.addToTail(new Node(3))
    linkedList.addToTail(new Node(4))
    linkedList.addToTail(new Node(5))
    linkedList.addToTail(new Node(6))
    console.log(`removed head: ${linkedList.removeFromHead()}`)
    console.log(`removed tail: ${linkedList.removeFromTail()}`)
    console.log(linkedList.toString())
}

main()
